ops/main.py

```diff
@@ -156,6 +156,73 @@ def _get_event_args(charm, bound_event):
     return [], {}
 
 
+class _JujuEvent:
+    """Encapsulate how to figure out what event juju wants us to run
+
+    Also knows how to run 'legacy' hooks when juju called us via dispatch.
+
+    Args:
+        charm_dir: the toplevel directory of the charm
+
+    Attributes:
+        name: the name of the event to run
+        is_dispatch: did juju call dispatch (as opposed to a hook symlink)
+    """
+    def __init__(self, charm_dir: Path):
+        self._charm_dir = charm_dir
+        self._exec_path = Path(sys.argv[0])
+        if self._exec_path.is_symlink() and self._exec_path.name != 'dispatch':
+            self._init_legacy()
+        else:
+            self._init_dispatch()
+
+    def run_any_legacy_hook(self):
+        if not self.is_dispatch:
+            # we *are* the legacy hook
+            return
+
+        dispatch_path = self._charm_dir / self._dispatch_path
+        if not dispatch_path.exists():
+            # nothing to do
+            return
+
+        if dispatch_path.resolve() == self._exec_path.resolve():
+            # we'd just be calling ourselves
```

Comment:
This handles dispatch being a symlink and the target being a symlink, but what about when this is a script and the target is a script? (reactive creates scripts instead of symlinks.)

Reply:
that's what's caught by 'OPERATOR_DISPATCH' (I didn't know reactive would trigger this)